# Exploring eBay Car Sales Data

1. Introduction
2. Cleaning Column Names
3. Initial Exploration and Cleaning
4. Exploring the odometer and price columns
5. Exploring the Data Columns
6. Dealing with incorrect registration year data
7. Exploring price by brand
8. Storing Aggregate data in a DataFrame
7. Further Analysis
8. Conclusion

# 1.Introduction

we'll work with a sampled 50,000 data points from the full dataset of used cars from eBay Kleinanzeigen, a classifieds section of the German eBay website.

The dataset was originally scraped and uploaded to Kaggle by user orgesleka.



The aim of this project is to clean the data and analyze the included used car listings

# Data Dictionary

-  dateCrawled - When this ad was first crawled. All field-values are taken from this date.
-  name - Name of the car.
-  offerType - The type of listing
-  price - The price on the ad to sell the car.
-  abtest - Whether the listing is included in an A/B test.
-  vehicleType - The vehicle Type.
-  yearOfRegistration - The year in which the car was first registered.
-  gearbox - The transmission type.
-  powerPS - The power of the car in PS.
-  model - The car model name.
-  odometer - How many kilometers the car has driven.
-  monthOfRegistration - The month in which the car was first registered.
-  fuelType - What type of fuel the car uses.
-  brand - The brand of the car.
-  notRepairedDamage - If the car has a damage which is not yet repaired.
-  dateCreated - The date on which the eBay listing was created.
-  nrOfPictures - The number of pictures in the ad.
-  postalCode - The postal code for the location of the vehicle.
-  lastSeenOnline - When the crawler saw this ad last online.

In [1]:
#import libraries
import pandas as pd
import numpy as np

In [2]:
#read data
autos=pd.read_csv('autos.csv',encoding='Latin-1')

In [3]:
autos

dateCrawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
...                    ...                                                ...   
49995  2016-03-27 14:38:19   Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon   
49996  2016-03-28 10:50:25  Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+...   
49997  2016-04-02 14:44:48                     Fiat_500_C_1.2_Dualogic_Lounge   
49998  2016-03-08 19:25:42                 Audi_A3_2.0_TDI_Sportback_Ambition   
49999  2016-03-14 00:42:12                                Opel_Vectra_1.6_16V   

       seller offerType    price   abtest vehicleType  yearOfRegistration  \
0      privat   Angebot   $5,000  control         bus                2004   
1      privat   Angebot   $8,500  control   limousine                1997   
2      privat   Angebot   $8,990     test   limousine                2009   
3      privat   Angebot   $4,350  control  kleinwagen                2007   
4      privat   Angebot   $1,350     test       kombi                2003   
...       ...       ...      ...      ...         ...                 ...   
49995  privat   Angebot  $24,900  control   limousine                2011   
49996  privat   Angebot   $1,980  control      cabrio                1996   
49997  privat   Angebot  $13,200     test      cabrio                2014   
49998  privat   Angebot  $22,900  control       kombi                2013   
49999  privat   Angebot   $1,250  control   limousine                1996   

         gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0        manuell      158  andere  150,000km                    3      lpg   
1      automatik      286     7er  150,000km                    6   benzin   
2        manuell      102    golf   70,000km                    7   benzin   
3      automatik       71  fortwo   70,000km                    6   benzin   
4        manuell        0   focus  150,000km                    7   benzin   
...          ...      ...     ...        ...                  ...      ...   
49995  automatik      239      q5  100,000km                    1   diesel   
49996    manuell       75   astra  150,000km                    5   benzin   
49997  automatik       69     500    5,000km                   11   benzin   
49998    manuell      150      a3   40,000km                   11   diesel   
49999    manuell      101  vectra  150,000km                    1   benzin   

            brand notRepairedDamage          dateCreated  nrOfPictures  \
0         peugeot              nein  2016-03-26 00:00:00             0   
1             bmw              nein  2016-04-04 00:00:00             0   
2      volkswagen              nein  2016-03-26 00:00:00             0   
3           smart              nein  2016-03-12 00:00:00             0   
4            ford              nein  2016-04-01 00:00:00             0   
...           ...               ...                  ...           ...   
49995        audi              nein  2016-03-27 00:00:00             0   
49996        opel              nein  2016-03-28 00:00:00             0   
49997        fiat              nein  2016-04-02 00:00:00             0   
49998        audi              nein  2016-03-08 00:00:00             0   
49999        opel              nein  2016-03-13 00:00:00             0   

       postalCode             lastSeen  
0           79588  2016-04-06 06:45:54  
1           71034  2016-04-06 14:45:08  
2           35394  2016-04-06 20:15:37  
3           33729  2016-03-15 03:16:28  
4           39218  2016-04-01 14:38:50  
...           ...                  ...  
4999

In [4]:
#print information about the autos
print(autos.info())
print(autos.head())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   dateCrawled          50000 non-null  object
 1   name                 50000 non-null  object
 2   seller               50000 non-null  object
 3   offerType            50000 non-null  object
 4   price                50000 non-null  object
 5   abtest               50000 non-null  object
 6   vehicleType          44905 non-null  object
 7   yearOfRegistration   50000 non-null  int64 
 8   gearbox              47320 non-null  object
 9   powerPS              50000 non-null  int64 
 10  model                47242 non-null  object
 11  odometer             50000 non-null  object
 12  monthOfRegistration  50000 non-null  int64 
 13  fuelType             45518 non-null  object
 14  brand                50000 non-null  object
 15  notRepairedDamage    40171 non-null  object
 16  date

In [5]:
#To print array of the columns
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

In [6]:
#edit column names from camelcase to snakecase
autos.columns = ['date_crawled', 'name', 'seller', 'offer_type', 'price', 'ab_test',
       'vehicle_type', 'registration_year', 'gear_box', 'power_ps', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'nr_of_pictures', 'postal_code',
       'last_seen']

autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price  ab_test vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

    gear_box  power_ps   model   odometer  registration_month fuel_type  \
0    manuell       158  andere  150,000km                   3       lpg   
1  automatik       286     7er  150,000km                   6    benzin   
2    manuell       102    golf   70,000km                   7    benzin   
3  automatik        71  fortwo   70,000km                   6    benzin   
4    manuell         0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  nr_of_pictures  \
0     peugeot              nein  2016-03-26 00:00:00               0   
1         bmw              nein  2016-04-04 00:00:00               0   
2  volkswagen              nein  2016-03-26 00:00:00               0   
3       smart              nein  2016-03-12 00:00:00               0   
4        ford              nein  2016-04-01 00:00:00               0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

# Initial Exploration and cleaning

Initially we will look for:

Text columns where all or almost all values are the same. These can often be dropped as they don't have useful information for analysis.
Examples of numeric data stored as text which can be cleaned and converted.

In [7]:
#To look at descriptive statistics for all columns.
autos.describe(include='all')

date_crawled         name  seller offer_type  price ab_test  \
count                 50000        50000   50000      50000  50000   50000   
unique                48213        38754       2          2   2357       2   
top     2016-03-11 22:38:16  Ford_Fiesta  privat    Angebot     $0    test   
freq                      3           78   49999      49999   1421   25756   
mean                    NaN          NaN     NaN        NaN    NaN     NaN   
std                     NaN          NaN     NaN        NaN    NaN     NaN   
min                     NaN          NaN     NaN        NaN    NaN     NaN   
25%                     NaN          NaN     NaN        NaN    NaN     NaN   
50%                     NaN          NaN     NaN        NaN    NaN     NaN   
75%                     NaN          NaN     NaN        NaN    NaN     NaN   
max                     NaN          NaN     NaN        NaN    NaN     NaN   

       vehicle_type  registration_year gear_box      power_ps  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  nr_of_pictures   postal_code            last_seen  
count                 50000         50000.0  50000.000000                50000  
unique                   76             NaN           NaN                39481  
top     2016-04-03 00:00:00             NaN           NaN  2016-04-07 06:17:27  
freq                   1946             NaN           NaN                    8  
mean                    NaN             0.0  50813.627300                  NaN  
std                     NaN             0.0  25779.747957                  NaN  
min                     NaN             0.0   1067.000000                  NaN  
25%                     NaN             0.0  30451.000000                  NaN  
50%                     NaN             0.0  49577.000000                  NaN  
75%                     NaN             0.0  71540.000000                  NaN  
max                     NaN             0.0  99998.000000                  NaN

- Any columns that have mostly one value that are candidates to be dropped
- Any columns that need more investigation.
- Any examples of numeric data stored as text that needs to be cleaned.

Found that the price and odometer columns are numeric values stored as text. For each column:
- Remove any non-numeric characters.
- Convert the column to a numeric dtype.
- Use DataFrame.rename() to rename the column to odometer_km.

In price column we obderved that the dollar sign followed by a number. we drop the dollar sign.

In [8]:
autos['price']

0         $5,000
1         $8,500
2         $8,990
3         $4,350
4         $1,350
          ...   
49995    $24,900
49996     $1,980
49997    $13,200
49998    $22,900
49999     $1,250
Name: price, Length: 50000, dtype: object

In [9]:
#cleaning price column
#convert 'object' to 'int' datatype and remove $ sign and ,
autos['price']=(autos['price']
                .str.replace(',','')
                .str.replace('$','')
                .astype(int))

In [10]:
autos['price']

0         5000
1         8500
2         8990
3         4350
4         1350
         ...  
49995    24900
49996     1980
49997    13200
49998    22900
49999     1250
Name: price, Length: 50000, dtype: int64

For odometer column,remove km and leave number,then rename the column.

In [11]:
# converting 'odometer' column from `object' to `int` datatype and remove non numerical characters
autos['odometer'] = (autos['odometer']
                            .str.replace('km','')
                            .str.replace(',','')
                            .astype(int)
                    )

In [12]:
# changing `odometer` column name to `odometer_in_km`
autos.rename({'odometer': 'odometer_in_km'}, axis=1, inplace=True)

In [13]:
#check after changes 
autos[['price','odometer_in_km']].head()

price  odometer_in_km
0   5000          150000
1   8500          150000
2   8990           70000
3   4350           70000
4   1350          150000

# Exploring the Odometer and Price Columns

Analyze the columns using minimum and maximum values and look for any values that look unrealistically high or low (outliers) that we might want to remove.

- Series.unique().shape to see how many unique values
- Series.describe() to view min/max/median/mean etc
- Series.value_counts(), with some variations:
- chained to .head() if there are lots of values.
- Because Series.value_counts() returns a series, we can use Series.sort_index() with ascending= True or False to view the highest and lowest values with their counts (can also chain to head() here).

In [14]:
autos['price'].unique().shape

(2357,)

There are 2357 unique values for the price values.

In [15]:
#To view statistics
autos["price"].describe()

count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64

Mean value is greater than the 50th percentile, median probably a skewed distribution. 



In [16]:
#To view highest and lowest value with counts
autos["price"].value_counts().sort_index(ascending = True).head(15)

0     1421
1      156
2        3
3        1
5        2
8        1
9        1
10       7
11       2
12       3
13       2
14       1
15       2
17       3
18       1
Name: price, dtype: int64

In [17]:
autos["price"].value_counts().sort_index(ascending = False).head(15)

99999999    1
27322222    1
12345678    3
11111111    2
10000000    1
3890000     1
1300000     1
1234566     1
999999      2
999990      1
350000      1
345000      1
299000      1
295000      1
265000      1
Name: price, dtype: int64

In [18]:
# dropping incorrect values and consider values between 100 and 350,000
autos = autos.loc[autos['price'].between(100,350000)]

In [19]:
autos["price"].value_counts().sort_index(ascending = False).head(20)

350000    1
345000    1
299000    1
295000    1
265000    1
259000    1
250000    1
220000    1
198000    1
197000    1
194000    1
190000    1
180000    1
175000    1
169999    1
169000    1
163991    1
163500    1
155000    1
151990    1
Name: price, dtype: int64

In [20]:
autos['odometer_in_km'].unique().shape

(13,)

In [21]:
autos['odometer_in_km'].value_counts().sort_index(ascending = False)

150000    31212
125000     5037
100000     2101
90000      1733
80000      1412
70000      1214
60000      1153
50000      1009
40000       814
30000       777
20000       757
10000       245
5000        760
Name: odometer_in_km, dtype: int64

odemeter reading columns looks good and no need to remove any values.

# Exploring the date columns

Right now, the date_crawled, last_seen, and ad_created columns are all identified as string values by pandas. Because these three columns are represented as strings, we need to convert the data into a numerical representation so we can understand it quantitatively. The other two columns are represented as numeric values, so we can use methods like Series.describe() to understand the distribution without any extra data processing.

In [22]:
autos[['date_crawled','ad_created','last_seen']][0:5]

date_crawled           ad_created            last_seen
0  2016-03-26 17:47:46  2016-03-26 00:00:00  2016-04-06 06:45:54
1  2016-04-04 13:38:56  2016-04-04 00:00:00  2016-04-06 14:45:08
2  2016-03-26 18:57:24  2016-03-26 00:00:00  2016-04-06 20:15:37
3  2016-03-12 16:58:10  2016-03-12 00:00:00  2016-03-15 03:16:28
4  2016-04-01 14:38:50  2016-04-01 00:00:00  2016-04-01 14:38:50

To select the first 10 characters in each column

In [23]:
date_crawled_date = autos["date_crawled"].str[:10].value_counts(normalize=True, dropna=False).sort_index()
ad_created_date = autos["ad_created"].str[:10].value_counts(normalize=True, dropna=False).sort_index()
last_seen_date = autos["last_seen"].str[:10].value_counts(normalize=True, dropna=False).sort_index()

In [24]:
date_crawled_date

2016-03-05    0.025361
2016-03-06    0.014039
2016-03-07    0.036061
2016-03-08    0.033179
2016-03-09    0.033013
2016-03-10    0.032287
2016-03-11    0.032598
2016-03-12    0.036911
2016-03-13    0.015677
2016-03-14    0.036662
2016-03-15    0.034319
2016-03-16    0.029467
2016-03-17    0.031499
2016-03-18    0.012898
2016-03-19    0.034734
2016-03-20    0.037803
2016-03-21    0.037201
2016-03-22    0.032888
2016-03-23    0.032287
2016-03-24    0.029446
2016-03-25    0.031499
2016-03-26    0.032308
2016-03-27    0.031126
2016-03-28    0.034962
2016-03-29    0.034112
2016-03-30    0.033738
2016-03-31    0.031851
2016-04-01    0.033697
2016-04-02    0.035605
2016-04-03    0.038611
2016-04-04    0.036538
2016-04-05    0.013064
2016-04-06    0.003173
2016-04-07    0.001389
Name: date_crawled, dtype: float64

In [25]:
date_crawled_date.describe()

count    34.000000
mean      0.029412
std       0.009766
min       0.001389
25%       0.029881
50%       0.032743
75%       0.034905
max       0.038611
Name: date_crawled, dtype: float64

To understand the distribution of year and month

In [26]:
print(autos[["registration_month", "registration_year"]].describe())

       registration_month  registration_year
count        48224.000000       48224.000000
mean             5.801634        2004.730964
std              3.676976          87.897388
min              0.000000        1000.000000
25%              3.000000        1999.000000
50%              6.000000        2004.000000
75%              9.000000        2008.000000
max             12.000000        9999.000000


In [27]:
autos["registration_year"].value_counts(normalize = True).sort_index()

1000    0.000021
1001    0.000021
1111    0.000021
1800    0.000041
1910    0.000041
          ...   
5911    0.000021
6200    0.000021
8888    0.000021
9000    0.000021
9999    0.000062
Name: registration_year, Length: 95, dtype: float64

In [28]:
autos["registration_month"].value_counts(normalize = True).sort_index()

0     0.089437
1     0.066316
2     0.060696
3     0.103289
4     0.083340
5     0.083278
6     0.088234
7     0.079670
8     0.064594
9     0.068804
10    0.074175
11    0.068534
12    0.069633
Name: registration_month, dtype: float64

Most of the cars are registered in the early 2000 and popular month is march.

# Dealing with Incorrect Registration Year Data

In registration_year column contains some odd values:

The minimum value is 1000, before cars were invented
The maximum value is 9999, many years into the future
Because a car can't be first registered after the listing was seen, any vehicle with a registration year above 2016 is definitely inaccurate. Determining the earliest valid year is more difficult. Realistically, it could be somewhere in the first few decades of the 1900s.

Let's count the number of listings with cars that fall outside the 1900 - 2016 interval and see if it's safe to remove those rows entirely, or if we need more custom logic.

We have issues in each column that could undermine the quality of our data:

For "registration_month", 10% of our values are invalid, since they are 0

For "registration_year", at least 4% of our values are also invalid, since it is impossible to list a car in 2016 that has been registered a year or more later. We also need to keep in mind those years that are too early.

We proceed to do as we did for the price. We will limit the years from 1900 (Ford T was invented 1908) to 2016, and the months from 1 to 12.

In [29]:
autos_clean = autos[(autos["registration_year"]>=1900) & (autos["registration_year"]<=2016)]


In [30]:
autos_clean.shape

(46352, 20)

In [31]:
# calculate the distribution of the registration year between 1900 - 2016

autos_clean["registration_year"].value_counts(normalize=True).sort_index()

1910    0.000043
1927    0.000022
1929    0.000022
1931    0.000022
1934    0.000043
          ...   
2012    0.028219
2013    0.017281
2014    0.014282
2015    0.008198
2016    0.025932
Name: registration_year, Length: 78, dtype: float64

In [32]:
autos_clean = autos[(autos["registration_month"]>=1) & (autos["registration_month"]<=12)]

In [33]:
# calculate the distribution of the registration month between 1 and 12
autos_clean["registration_month"].value_counts(normalize=True).sort_index()

1     0.072829
2     0.066658
3     0.113434
4     0.091526
5     0.091458
6     0.096901
7     0.087495
8     0.070939
9     0.075562
10    0.081460
11    0.075266
12    0.076473
Name: registration_month, dtype: float64

# Exploring Price by Brand

Explore the unique values in the brand column, and decide on which brands you want to aggregate by.
You might want to select the top 20, or you might want to select those that have over a certain percentage of the total values (e.g. > 5%).

In [34]:
# identify the unique brands in our dataset and their counts

autos_clean["brand"].value_counts().sort_values(ascending=False)

volkswagen        9210
bmw               4875
opel              4588
mercedes_benz     4344
audi              3845
ford              3016
renault           2060
peugeot           1299
fiat              1122
seat               835
skoda              739
nissan             686
mazda              654
smart              641
citroen            612
toyota             579
hyundai            448
volvo              405
mini               404
sonstige_autos     399
mitsubishi         353
honda              351
kia                332
alfa_romeo         289
porsche            272
suzuki             262
chevrolet          252
chrysler           151
dacia              121
daihatsu           103
jeep               102
land_rover          93
subaru              86
saab                75
jaguar              68
daewoo              68
rover               58
lancia              47
trabant             43
lada                24
Name: brand, dtype: int64

In [35]:
# select top 20 brands
top_20_brands = autos_clean["brand"].value_counts().sort_values(ascending=False).index[:20]

In [36]:
top_20_brands

Index(['volkswagen', 'bmw', 'opel', 'mercedes_benz', 'audi', 'ford', 'renault',
       'peugeot', 'fiat', 'seat', 'skoda', 'nissan', 'mazda', 'smart',
       'citroen', 'toyota', 'hyundai', 'volvo', 'mini', 'sonstige_autos'],
      dtype='object')

In [37]:
#Create an empty dictionary to hold your aggregate data
top_brands_price={}
#Loop over your selected brands, and assign the mean price to the dictionary
for brand in top_20_brands:
    #select only top brands from autos clean column
    rows_selected=autos_clean[autos_clean["brand"] == brand]
    #calculate mean price value
    mean_price=rows_selected['price'].mean()
    #assign the mean price to the dictionary as key
    top_brands_price[brand]=mean_price

In [38]:
top_brands_price

{'volkswagen': 5707.066992399566,
 'bmw': 8621.819282051281,
 'opel': 3168.253923278117,
 'mercedes_benz': 8860.345073664825,
 'audi': 9698.319375812744,
 'ford': 3955.9436339522545,
 'renault': 2594.3616504854367,
 'peugeot': 3216.1993841416474,
 'fiat': 2960.4732620320856,
 'seat': 4515.700598802395,
 'skoda': 6568.420838971583,
 'nissan': 4879.304664723032,
 'mazda': 4322.5749235474,
 'smart': 3651.6209048361934,
 'citroen': 3897.87908496732,
 'toyota': 5277.101899827288,
 'hyundai': 5624.618303571428,
 'volvo': 5048.106172839506,
 'mini': 10722.965346534653,
 'sonstige_autos': 13846.40350877193}

In [39]:
#proper formatting
for key, value in top_brands_price.items():
    print("{} : ${:,.2f}".format(key, value))

volkswagen : $5,707.07
bmw : $8,621.82
opel : $3,168.25
mercedes_benz : $8,860.35
audi : $9,698.32
ford : $3,955.94
renault : $2,594.36
peugeot : $3,216.20
fiat : $2,960.47
seat : $4,515.70
skoda : $6,568.42
nissan : $4,879.30
mazda : $4,322.57
smart : $3,651.62
citroen : $3,897.88
toyota : $5,277.10
hyundai : $5,624.62
volvo : $5,048.11
mini : $10,722.97
sonstige_autos : $13,846.40


The brand with most expensive is audi and least expensive renault.
We observed that in the top 6 brands, there's a distinct price gap.
Audi, BMW and Mercedes Benz are more expensive
Ford and Opel are less expensive
Volkswagen is in between

# Storing Aggregate Data in a DataFrame

For the top 6 brands, let's use aggregation to understand the average mileage for those cars and if there's any visible link with mean price. While our natural instinct may be to display both aggregated series objects and visually compare them, this has a few limitations:

it's difficult to compare more than two aggregate series objects if we want to extend to more columns
- we can't compare more than a few rows from each series object
- we can only sort by the index (brand name) of both series objects so we can easily make visual comparisons

In [40]:
bmp_series=pd.Series(top_brands_price)
print(bmp_series)

volkswagen         5707.066992
bmw                8621.819282
opel               3168.253923
mercedes_benz      8860.345074
audi               9698.319376
ford               3955.943634
renault            2594.361650
peugeot            3216.199384
fiat               2960.473262
seat               4515.700599
skoda              6568.420839
nissan             4879.304665
mazda              4322.574924
smart              3651.620905
citroen            3897.879085
toyota             5277.101900
hyundai            5624.618304
volvo              5048.106173
mini              10722.965347
sonstige_autos    13846.403509
dtype: float64


In [41]:
df = pd.DataFrame(bmp_series, columns=['mean_price'])
df

mean_price
volkswagen       5707.066992
bmw              8621.819282
opel             3168.253923
mercedes_benz    8860.345074
audi             9698.319376
ford             3955.943634
renault          2594.361650
peugeot          3216.199384
fiat             2960.473262
seat             4515.700599
skoda            6568.420839
nissan           4879.304665
mazda            4322.574924
smart            3651.620905
citroen          3897.879085
toyota           5277.101900
hyundai          5624.618304
volvo            5048.106173
mini            10722.965347
sonstige_autos  13846.403509

In [42]:
autos_clean

date_crawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
...                    ...                                                ...   
49995  2016-03-27 14:38:19   Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon   
49996  2016-03-28 10:50:25  Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+...   
49997  2016-04-02 14:44:48                     Fiat_500_C_1.2_Dualogic_Lounge   
49998  2016-03-08 19:25:42                 Audi_A3_2.0_TDI_Sportback_Ambition   
49999  2016-03-14 00:42:12                                Opel_Vectra_1.6_16V   

       seller offer_type  price  ab_test vehicle_type  registration_year  \
0      privat    Angebot   5000  control          bus               2004   
1      privat    Angebot   8500  control    limousine               1997   
2      privat    Angebot   8990     test    limousine               2009   
3      privat    Angebot   4350  control   kleinwagen               2007   
4      privat    Angebot   1350     test        kombi               2003   
...       ...        ...    ...      ...          ...                ...   
49995  privat    Angebot  24900  control    limousine               2011   
49996  privat    Angebot   1980  control       cabrio               1996   
49997  privat    Angebot  13200     test       cabrio               2014   
49998  privat    Angebot  22900  control        kombi               2013   
49999  privat    Angebot   1250  control    limousine               1996   

        gear_box  power_ps   model  odometer_in_km  registration_month  \
0        manuell       158  andere          150000                   3   
1      automatik       286     7er          150000                   6   
2        manuell       102    golf           70000                   7   
3      automatik        71  fortwo           70000                   6   
4        manuell         0   focus          150000                   7   
...          ...       ...     ...             ...                 ...   
49995  automatik       239      q5          100000                   1   
49996    manuell        75   astra          150000                   5   
49997  automatik        69     500            5000                  11   
49998    manuell       150      a3           40000                  11   
49999    manuell       101  vectra          150000                   1   

      fuel_type       brand unrepaired_damage           ad_created  \
0           lpg     peugeot              nein  2016-03-26 00:00:00   
1        benzin         bmw              nein  2016-04-04 00:00:00   
2        benzin  volkswagen              nein  2016-03-26 00:00:00   
3        benzin       smart              nein  2016-03-12 00:00:00   
4        benzin        ford              nein  2016-04-01 00:00:00   
...         ...         ...               ...                  ...   
49995    diesel        audi              nein  2016-03-27 00:00:00   
49996    benzin        opel              nein  2016-03-28 00:00:00   
49997    benzin        fiat              nein  2016-04-02 00:00:00   
49998    diesel        audi              nein  2016-03-08 00:00:00   
49999    benzin        opel              nein  2016-03-13 00:00:00   

       nr_of_pictures  postal_code            last_seen  
0                   0        79588  2016-04-06 06:45:54  
1                   0        71034  2016-04-06 14:45:08  
2                   0        35394  2016-04-06 20:15:37  
3                   0        33729  2016-03-15 03:16:28  
4                   0        39218  2016-04-01 14:38:50  
...               ...          ...                

In [43]:
#Create an empty dictionary to hold your aggregate data
brand_mean_mileage={}
#Loop over your selected brands, and assign the mean mileage to the dictionary
for brand in top_20_brands:
    #select only top brands from autos clean column
    rows_selected=autos_clean[autos_clean["brand"] == brand]
    #calculate mean price value
    mean_mileage=rows_selected['odometer_in_km'].mean()
    #assign the mean price to the dictionary as key
    brand_mean_mileage[brand]=mean_mileage

In [44]:
bmm_series = pd.Series(brand_mean_mileage)

In [45]:
df['mean_mileage']=bmm_series

In [46]:
df.sort_values("mean_price", ascending=False)

mean_price   mean_mileage
sonstige_autos  13846.403509   91817.042607
mini            10722.965347   88551.980198
audi             9698.319376  128908.972692
mercedes_benz    8860.345074  130965.699816
bmw              8621.819282  132591.794872
skoda            6568.420839  110250.338295
volkswagen       5707.066992  128361.563518
hyundai          5624.618304  105647.321429
toyota           5277.101900  115500.863558
volvo            5048.106173  138851.851852
nissan           4879.304665  117543.731778
seat             4515.700599  121263.473054
mazda            4322.574924  123700.305810
ford             3955.943634  123922.413793
citroen          3897.879085  118921.568627
smart            3651.620905   99656.786271
peugeot          3216.199384  126397.228637
opel             3168.253923  128833.914560
fiat             2960.473262  116421.568627
renault          2594.361650  128101.941748

Higher brand has the most mileage

# Find the most common brand/model combinations

In [47]:
autos_by_brandmodel = autos_clean.groupby(["brand", "model"])["brand"].count()

In [48]:
autos_by_brandmodel

brand       model   
alfa_romeo  145          4
            147         83
            156         76
            159         30
            andere      54
                        ..
volvo       v40         86
            v50         28
            v60          3
            v70         84
            xc_reihe    47
Name: brand, Length: 291, dtype: int64

# Further Analysis

In [50]:
unrepaired_damage_map = {
    'nein': 'no',
    'ja': 'yes'
}

In [51]:
autos_clean['unrepaired_damage'] = autos_clean['unrepaired_damage'].map(unrepaired_damage_map)

<ipython-input-51-95dee44114f8>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  autos_clean['unrepaired_damage'] = autos_clean['unrepaired_damage'].map(unrepaired_damage_map)


In [52]:
# Calculate the average price of damaged cars
damaged_cars_avg_price = autos_clean[autos_clean['unrepaired_damage'] == 'yes']['price'].mean()

# Calculate the average price of non-damaged cars
non_damaged_cars_avg_price = autos_clean[autos_clean['unrepaired_damage'] == 'no']['price'].mean()

# Calculate the price difference
price_difference = non_damaged_cars_avg_price - damaged_cars_avg_price

print("Average price of damaged cars: {:.2f}".format(damaged_cars_avg_price))
print("Average price of non-damaged cars: {:.2f}".format(non_damaged_cars_avg_price))
print("Price difference: {:.2f}".format(price_difference))

Average price of damaged cars: 2413.93
Average price of non-damaged cars: 7238.68
Price difference: 4824.75
